src/server/pfs/server/trigger.go

```diff
@@ -16,86 +15,73 @@ import (
 )
 
 // triggerCommit is called when a commit is finished, it updates branches in
-// the repo if they trigger on the change and returns all branches which were
-// moved by this call.
+// the repo if they trigger on the change
 func (d *driver) triggerCommit(
 	txnCtx *txncontext.TransactionContext,
 	commit *pfs.Commit,
-) ([]*pfs.Branch, error) {
+) error {
 	repoInfo := &pfs.RepoInfo{}
 	if err := d.repos.ReadWrite(txnCtx.SqlTx).Get(pfsdb.RepoKey(commit.Branch.Repo), repoInfo); err != nil {
-		return nil, err
-	}
-	newHead := &pfs.CommitInfo{}
-	if err := d.commits.ReadWrite(txnCtx.SqlTx).Get(pfsdb.CommitKey(commit), newHead); err != nil {
-		return nil, err
-	}
-	// find which branches this commit is the head of
-	headBranches := make(map[string]bool)
-	// The commit _was_ the branch head at some point in time, although it is
-	// not guaranteed to still be the branch head. This can happen on a
-	// downstream pipeline with triggers - the upstream pipeline may have
-	// multiple unfinished commits in its output branch that will be finished
-	// one at a time. Without this code, only finishing the _last_ commit would
-	// have a chance of triggering.
-	headBranches[newHead.Commit.Branch.Name] = true
-	for _, b := range repoInfo.Branches {
-		bi := &pfs.BranchInfo{}
-		if err := d.branches.ReadWrite(txnCtx.SqlTx).Get(pfsdb.BranchKey(b), bi); err != nil {
-			return nil, err
-		}
-		if bi.Head != nil && bi.Head.ID == commit.ID {
-			headBranches[b.Name] = true
-		}
+		return err
 	}
-	triggeredBranches := map[string]bool{}
-	var result []*pfs.Branch
-	var triggerBranch func(branch *pfs.Branch) error
-	triggerBranch = func(branch *pfs.Branch) error {
-		if triggeredBranches[branch.Name] {
-			return nil
+	commitInfo := &pfs.CommitInfo{}
+	if err := d.commits.ReadWrite(txnCtx.SqlTx).Get(pfsdb.CommitKey(commit), commitInfo); err != nil {
+		return err
+	}
+
+	// Track any branches that are triggered and their new (alias) head commit
+	triggeredBranches := map[string]*pfs.CommitInfo{}
+	triggeredBranches[commitInfo.Commit.Branch.Name] = commitInfo
+
+	var triggerBranch func(branch *pfs.Branch) (*pfs.CommitInfo, error)
+	triggerBranch = func(branch *pfs.Branch) (*pfs.CommitInfo, error) {
+		if newHead, ok := triggeredBranches[branch.Name]; ok {
+			return newHead, nil
 		}
-		triggeredBranches[branch.Name] = true
+
 		bi := &pfs.BranchInfo{}
 		if err := d.branches.ReadWrite(txnCtx.SqlTx).Get(pfsdb.BranchKey(branch), bi); err != nil {
-			return err
+			return nil, err
 		}
+
+		triggeredBranches[branch.Name] = nil
```

Comment:
Isn't this necessarily nil if you reach here?

Reply:
It would be `nil` if you did a read, but it isn't present in the map - this adds an entry to mark it as processed.